### DISCUSSION TITLE
Programmatic front matter

### DISCUSSION THREAD

**Discussion by dominikh** [kaushalmodi/ox-hugo#697] [Ideas]
Would it be possible to have a hook or callback that let's us insert custom Hugo front matter at the time of export? Concretely, I want to set `lastmod` to the modification time of the Org file (using "one post per Org file") without having to materialize it in the Org file's front matter.

Such a hook could be used for other computable front matter, too. For example, one could set the section based on the file name.

One possible API would be a configurable callback, that takes the Org buffer as input and returns a list of key/value pairs of Hugo front matter to insert.

**Comment by kaushalmodi** [comment]
There's isn't a generic solution. 

But specific to your usecase, there something already available: `#+hugo_auto_set_lastmod: t`.

Search for "auto_set" on https://ox-hugo.scripter.co/doc/dates/.

**Comment by dominikh** [comment]
> #+hugo_auto_set_lastmod: t

Doesn't that set lastmod to the current time? I want to set it to the file's mtime. I export all my notes in batch, not only those that have changed. So with `auto_set_lastmod`, all posts have the current time as the lastmod time.

> There's isn't a generic solution.

I know, I was wondering if one should be added.

**Reply by kaushalmodi** [reply]
> I want to set it to the file's mtime. 

That would work only for file based export flow. 

> I was wondering if one should be added.

I would accept a pull request if the feature is complete with documentation and tests.

**Comment by kaushalmodi** [comment]
This works.

## Subtree based flow

```org
#+hugo_base_dir: .

* Post
:PROPERTIES:
:EXPORT_FILE_NAME: foo
:EXPORT_HUGO_CUSTOM_FRONT_MATTER: :lastmod (format-time-string "%Y-%m-%dT%T%z" (nth 5 (file-attributes (buffer-file-name))))
:END:
Hello
```

## File based flow

```org
#+hugo_base_dir: .

#+title: Post
#+export_file_name: foo

#+hugo_custom_front_matter: :lastmod (format-time-string "%Y-%m-%dT%T%z" (nth 5 (file-attributes (buffer-file-name))))

Hello
```

**Reply by kaushalmodi** [reply]
@dominikh To make the above a bit succinct, you can choose to define your own defun called `mtime` that calls `(format-time-string "%Y-%m-%dT%T%z" (nth 5 (file-attributes (buffer-file-name))))` and then do `#+hugo_custom_front_matter: :lastmod (mtime)`.

**Reply by dominikh** [reply]
I wasn't aware hugo_custom_front_matter could contain lisp expressions, that's great. 

Now, if one could set that as a variable (e.g. via `let`), then that'd be perfect; similar to how you can set `org-hugo-auto-set-lastmod`. That way, it could go in .dir-locals.el and work for all notes automatically.

**Comment by kaushalmodi** [comment]
This is also possible:

## Org setup file `mysetup.org`

```org
#+hugo_custom_front_matter: :lastmod (format-time-string "%Y-%m-%dT%T%z" (nth 5 (file-attributes (buffer-file-name))))
```

## Post file `mypost.org`

```org
#+hugo_base_dir: .
#+setupfile: mysetup.org

#+title: Post
#+export_file_name: foo

Hello
```

Then, you only need `#+setupfile: mysetup.org` in all your post files, and you can load up `mysetup.org` will all the elisp-based front-matter you need 😄 .

**Reply by dominikh** [reply]
Thank you, I'll play around with that.

**Comment by dominikh** [comment]
I've arrived at the following fully programatic way that doesn't require any changes to the Org files themselves:

```
(defun dh/org-export-before-parsing (backend)
  (when (string= backend "hugo")
	(org-roam-set-keyword
	 "hugo_lastmod"
	 (format-time-string "%Y-%m-%d" (file-attribute-modification-time (file-attributes (buffer-file-name)))))))

(add-to-list 'org-export-before-parsing-functions #'dh/org-export-before-parsing)
```
